microjson/src/json.rs

```diff
@@ -0,0 +1,293 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use collections::string::String;
+use core::convert::From;
+use core::str;
+
+#[derive(Debug, PartialEq)]
+pub enum JsonToken {
+    StartObject,
+    EndObject,
+    StartArray,
+    EndArray,
+    PropertyName(String),
+    Literal(String),
+    Null,
+    True,
+    False,
+    Done,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum JsonError {
+    UnexpectedCharacter,
+    UnexpecteEof,
+    InvalidString,
+}
+
+impl From<JsonError> for () {
+    fn from(_: JsonError) -> () {
+        ()
+    }
+}
+
+
+#[derive(Debug, PartialEq)]
+enum TokenizerState {
+    Start,
+    ExpectProperty,
+    ExpectValue,
+    AfterArray,
+    InArray,
+}
+
+pub struct JsonTokenizer<'a> {
+    buffer: &'a [u8],
+    len: usize,
+    pos: usize,
+    state: TokenizerState,
+    depth: usize,
+}
+
+macro_rules! error_if_eof {
+    ($s:ident) => (
+        if $s.eof() {
+            return Err(JsonError::UnexpecteEof);
+        }
+    )
+}
+
+impl<'a> JsonTokenizer<'a> {
+    pub fn new(text: &'a str) -> Self {
+        let buffer = text.as_bytes();
+        JsonTokenizer {
+            buffer: buffer,
+            len: buffer.len(),
+            pos: 0,
+            state: TokenizerState::Start,
+            depth: 0,
+        }
+    }
+
+    fn eof(&self) -> bool {
+        self.pos >= self.len
+    }
+
+    fn next(&mut self) -> Result<u8, JsonError> {
+        let c = self.peek()?;
+        self.pos += 1;
+        Ok(c)
+    }
+
+    fn peek(&mut self) -> Result<u8, JsonError> {
+        if self.eof() {
+            return Err(JsonError::UnexpecteEof);
+        }
+        let c = self.buffer[self.pos];
+        Ok(c)
+    }
+
+    fn eat_ws(&mut self) {
+        loop {
+            if let Ok(c) = self.next() {
```

Comment:
Now that I see this rewritten, it could be further simplified to use a while let